torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1305,7 +1305,24 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 if self.gradient_postdivide_factor > 1:
                     # Average grad by world_size for consistency with PyTorch DDP.
                     output.div_(self.gradient_postdivide_factor)
-                param.grad.data = output
+                # To support gradient accumulation without `no_sync()`, we save
```

Comment:
looks like it is important to do the following before flatten param.grad and pass it to reduce_scatter_base, as commented in https://github.com/facebookresearch/fairscale/commit/4fa2ab9b70d55f4e3e94239465b877c9f3733037

grad = param.grad.data
param.grad = None

This is to avoid race among multiple re-entrant backward passes.  for multiple re-entrant backward passes, e.g., it is possible that the second backward pass computation is ahead of first backward pass reduction, as reduction is in another stream and is async. In this case, first backward pass may be reducing the second pass's param.grad, which is not correct. 

so it is better to reduce_scatter the saved local grad and accumulate the reduced grad to saved_grad_shard. And then in the final callback of main backward pass, let param.grad point to saved_grad_shard. 

let's add change here and add comments as well